### PR TITLE
gcc: fix build Darwin builds on staging-next

### PR DESCRIPTION
**Pull request by reckenrode**
## Description of changes

This PR fixes the build of all versions of GCC on Darwin in staging-next #263535 except for GCC 4.8. GCC 4.8 hasn’t built successfully for almost a decade, so the PR just disables it.

### aarch64-darwin
* GCC 11: https://hydra.nixos.org/build/239466277

### x86_64-darwin
* GCC 10: https://hydra.nixos.org/build/239551156
* GCC 9: https://hydra.nixos.org/build/239612904
* GCC 8: https://hydra.nixos.org/build/239628060
* GCC 7: https://hydra.nixos.org/build/239554679
* GCC 6: https://hydra.nixos.org/build/239525116

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by reckenrode** [comment]
I updated the PR to clean it up and avoid unnecessary rebuilds (especially of GCC 12 and 13). I also fixed the merge conflict with #265934.

File: pkgs/development/compilers/gcc/default.nix [hunk 1]
```diff
@@ -408,10 +409,24 @@ lib.pipe ((callFile ./common/builder.nix {}) ({
       maintainers
     ;
   } // lib.optionalAttrs (!atLeast11) {
-    badPlatforms = if !is49 then [ "aarch64-darwin" ] else lib.platforms.darwin;
+    badPlatforms = if !(is48 || is49) then [ "aarch64-darwin" ] else lib.platforms.darwin;
   };
-} // optionalAttrs is7 {
-  env.NIX_CFLAGS_COMPILE = lib.optionalString (stdenv.cc.isClang && langFortran) "-Wno-unused-command-line-argument";
+} // optionalAttrs (!atLeast11 && stdenv.cc.isClang) {
+  NIX_CFLAGS_COMPILE = toString (
```
[the remark this answers is not in the document]
Comment (reckenrode) [reply pkgs/development/compilers/gcc/default.nix:415]: That doesn’t actually work for some reason. It doesn’t pick up the `NIX_CFLAGS_COMPILE` when you run the build if it it’s in `env`.

[the remark this answers is not in the document]
Comment (reckenrode) [reply pkgs/development/compilers/gcc/default.nix:415]: Would it help to add a comment explaining that it doesn’t work using `env.NIX_CFLAGS_COMPILE`?

[the remark this answers is not in the document]
Comment (wegank) [reply pkgs/development/compilers/gcc/default.nix:415]: Adding `env.` or not doesn't change the value of `NIX_CFLAGS_COMPILE` when I do `nix build` on my side, though. Is this due to different behaviours in `nix develop`?

[the remark this answers is not in the document]
Comment (reckenrode) [reply pkgs/development/compilers/gcc/default.nix:415]: That’s strange. If I look at the contents of the raw derivation, `NIX_CFLAGS_COMPILE` is only present when it’s at the top level (without `env`). That’s not the case with other derivations. I don’t know what’s happening with GCC.

[the remark this answers is not in the document]
Comment (reckenrode) [reply pkgs/development/compilers/gcc/default.nix:415]: How are you viewing the variables in `nix build`?

[the remark this answers is not in the document]
Comment (reckenrode) [reply pkgs/development/compilers/gcc/default.nix:415]: If it matters, I’m using Nix 2.17.1. nixpkgs is staging-next.

```console
$ nix-info -m
 - system: `"aarch64-darwin"`
 - host os: `Darwin 23.1.0, macOS 14.1`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.17.1`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/etc/nix/inputs/nixpkgs`
```

[the remark this answers is not in the document]
Comment (wegank) [reply pkgs/development/compilers/gcc/default.nix:415]: By adding `echo $NIX_CFLAGS_COMPILE && exit 1` in certain phases, for example `preConfigure`.

[the remark this answers is not in the document]
Comment (reckenrode) [reply pkgs/development/compilers/gcc/default.nix:415]: Ah. I’ll add and push an update once I do a local build. Should I also add it to the `NIX_LDFLAGS` later in that file while I’m at it?

[the remark this answers is not in the document]
Comment (wegank) [reply pkgs/development/compilers/gcc/default.nix:415]: I don't think it will trigger any rebuilds, in which case, yes, please.

[the remark this answers is not in the document]
Comment (reckenrode) [reply pkgs/development/compilers/gcc/default.nix]: The builds finished without issue. Only GCC 6–10 rebuilt, which was expected. 11–13 did not rebuild.

**Comment by amjoseph-nixpkgs** [comment]
This shouldn't cause any rebuilds on Linux, should it?

File: pkgs/development/compilers/gcc/all.nix [hunk 2]
```diff
@@ -34,9 +34,6 @@ let
               else            /* "4.8" */   isl_0_14;
       } // lib.optionalAttrs (majorMinorVersion == "4.8") {
         texinfo = texinfo5; # doesn't validate since 6.1 -> 6.3 bump
-      } // lib.optionalAttrs (majorMinorVersion == "4.9") {
-        # Build fails on Darwin with clang
-        stdenv = if stdenv.isDarwin then gccStdenv else stdenv;
```
Comment (amjoseph-nixpkgs) [inline pkgs/development/compilers/gcc/all.nix:39]: Please move this to another PR; it's an easy approval.  Also, that way it can go directly to `master`.

File: pkgs/development/compilers/gcc/default.nix [hunk 3]
```diff
@@ -408,14 +409,28 @@ lib.pipe ((callFile ./common/builder.nix {}) ({
       maintainers
     ;
   } // lib.optionalAttrs (!atLeast11) {
-    badPlatforms = if !is49 then [ "aarch64-darwin" ] else lib.platforms.darwin;
+    badPlatforms = if !(is48 || is49) then [ "aarch64-darwin" ] else lib.platforms.darwin;
```
Comment (amjoseph-nixpkgs) [inline pkgs/development/compilers/gcc/default.nix:412]: Likewise.

**Comment by reckenrode** [comment]
> This shouldn't cause any rebuilds on Linux, should it?

According to ofborg, there are eighteen rebuilds on Linux.

File: pkgs/development/compilers/gcc/default.nix [hunk 4]
```diff
@@ -408,14 +409,28 @@ lib.pipe ((callFile ./common/builder.nix {}) ({
...
   };
-} // optionalAttrs is7 {
-  env.NIX_CFLAGS_COMPILE = lib.optionalString (stdenv.cc.isClang && langFortran) "-Wno-unused-command-line-argument";
```
Comment (amjoseph-nixpkgs) [inline pkgs/development/compilers/gcc/default.nix:414]: Why was this deleted?

File: pkgs/development/compilers/gcc/default.nix [hunk 5]
```diff
@@ -408,14 +409,28 @@ lib.pipe ((callFile ./common/builder.nix {}) ({
...
+} // lib.optionalAttrs (!atLeast10 && stdenv.hostPlatform.isDarwin) {
+  # GCC <10 requires default cctools `strip` instead of `llvm-strip` used by Darwin bintools.
+  preBuild = ''
+    makeFlagsArray+=('STRIP=${lib.getBin darwin.cctools-port}/bin/${stdenv.cc.targetPrefix}strip')
+  '';
 } // optionalAttrs (!atLeast7) {
   env.langJava = langJava;
 } // optionalAttrs atLeast6 {
-  NIX_LDFLAGS = lib.optionalString  hostPlatform.isSunOS "-lm";
+  env.NIX_LDFLAGS = lib.optionalString hostPlatform.isSunOS "-lm";
```
Comment (amjoseph-nixpkgs) [inline pkgs/development/compilers/gcc/default.nix:433]: This change might be a good idea, but it doesn't need to go to `staging-next`.  It also doesn't belong in a Darwin-specific PR.

File: pkgs/development/compilers/gcc/default.nix [hunk 6]
```diff
@@ -408,14 +409,28 @@ lib.pipe ((callFile ./common/builder.nix {}) ({
...
+  env.NIX_CFLAGS_COMPILE = toString (
+    lib.optionals (lib.versionAtLeast (lib.getVersion stdenv.cc) "16") [
+      # Silence errors related to use of implicitly declared library functions and implicit ints,
+      # which can cause unwanted `configure` feature detection failures.
+      "-Wno-error=implicit-function-declaration"
+      "-Wno-error=implicit-int"
+    ] ++ lib.optionals (is7 && langFortran) [
+      "-Wno-unused-command-line-argument"
+    ]
+  );
```
Comment (amjoseph-nixpkgs) [inline pkgs/development/compilers/gcc/default.nix:424]: This change isn't Darwin-specific; please move it to a separate PR.

**Review by ghost — CHANGES_REQUESTED** [review]
There is **way** too much going on at once in this PR.  Much of it is not Darwin-specific, so it doesn't belong in a PR titled "fix Darwin builds".  A bunch of stuff in here has no justification for going directly to `staging-next`, hours/days before the very last pre-branchoff merge of `staging-next` into staging.

Come on, this is not the right way to do this.

File: pkgs/development/compilers/gcc/default.nix (hunk 5 above)
Comment (amjoseph-nixpkgs) [reply pkgs/development/compilers/gcc/default.nix:433]: See
- https://github.com/NixOS/nixpkgs/pull/266193

Comment (reckenrode) [reply pkgs/development/compilers/gcc/default.nix:433]: This was requested by @wegank. If there’s already work being done on it, I’ll just drop the commit.

File: pkgs/development/compilers/gcc/default.nix (hunk 4 above)
Comment (reckenrode) [reply pkgs/development/compilers/gcc/default.nix:414]: It was consolidated with the other change to `NIX_CFLAGS_COMPILE`.

**Comment by amjoseph-nixpkgs** [comment]
> Edit: I think it’s due to the way `NIX_CFLAGS_COMPILE` was changed.

It's due to this:

- https://github.com/NixOS/nixpkgs/pull/265935#discussion_r1386007319

I still don't understand why that was deleted if this emergency PR is just trying to fix Darwin.

File: pkgs/development/compilers/gcc/all.nix (hunk 2 above)
Comment (reckenrode) [reply pkgs/development/compilers/gcc/all.nix:39]: Moved to https://github.com/NixOS/nixpkgs/pull/266194.

File: pkgs/development/compilers/gcc/patches/default.nix [hunk 7]
```diff
@@ -191,19 +194,30 @@ in
   sha256 = "sha256-XtykrPd5h/tsnjY1wGjzSOJ+AyyNLsfnjuOZ5Ryq9vA=";
 })
 
+# Fix undefined symbol errors when building older versions with clang
+++ optional (!atLeast11 && stdenv.cc.isClang) ./clang-genconditions.patch
```
Comment (amjoseph-nixpkgs) [inline pkgs/development/compilers/gcc/patches/default.nix:198]:
```suggestion
++ optional (!atLeast11 && stdenv.cc.isClang && stdenv.hostPlatform.isDarwin) ./clang-genconditions.patch
```

File: pkgs/development/compilers/gcc/patches/default.nix [hunk 8]
```diff
@@ -284,6 +298,11 @@ in
   })
 ]
 
+# The clang-based assembler used in darwin.cctools-llvm (LLVM >11) does not support piping input.
+# Fortunately, it does not exhibit the problem GCC has with the cctools assembler.
+# This patch can be dropped should darwin.cctools-llvm ever implement support.
+++ optional (!atLeast7 && hostPlatform.isDarwin && lib.versionAtLeast (lib.getVersion stdenv.cc) "12") ./4.9/darwin-clang-as.patch
```
Comment (amjoseph-nixpkgs) [inline pkgs/development/compilers/gcc/patches/default.nix:304]:
```suggestion
++ optional (!atLeast6 && hostPlatform.isDarwin && lib.versionAtLeast (lib.getVersion stdenv.cc) "12") ./4.9/darwin-clang-as.patch
```

File: pkgs/development/compilers/gcc/patches/default.nix [hunk 9]
```diff
@@ -284,6 +298,11 @@ in
...
+
```
Comment (amjoseph-nixpkgs) [inline pkgs/development/compilers/gcc/patches/default.nix:305]: The changes to `gcc/patches/default.nix` and the patches themselves look okay; if you apply the two minor changes I suggested above and put them in a separate PR targeted to `master` I will approve+merge it.  Then we can focus on the stuff in `default.nix`.

File: pkgs/development/compilers/gcc/patches/default.nix (hunk 8 above)
Comment (amjoseph-nixpkgs) [reply pkgs/development/compilers/gcc/patches/default.nix:304]: (or else move it to the "gcc 6 and earlier" section if you want it applied to `gcc6`)

**Comment by reckenrode** [comment]
> I still don't understand why that was deleted if this emergency PR is just trying to fix Darwin.

It wasn’t deleted. It was moved. `//` doesn’t merge attrsets. If I added the condition for clang 16 after the existing one, `NIX_CFLAGS_COMPILE` wouldn’t have `-Wno-unused-command-line-argument` when building gfortran7 on a clang 16 platform like Darwin.

File: pkgs/development/compilers/gcc/patches/default.nix (hunk 8 above)
Comment (reckenrode) [reply pkgs/development/compilers/gcc/patches/default.nix:304]: It should apply to GCC 6, so I’ll move it to the correct section.

File: pkgs/development/compilers/gcc/patches/default.nix (hunk 7 above)
Comment (reckenrode) [reply pkgs/development/compilers/gcc/patches/default.nix:198]: This technically isn’t a Darwin-specific problem even if it’s primarily affecting x86_64-darwin. There’s an issue in GCC’s Bugzilla regarding the issue. It’s mentioned already in the commit message and at the top of the patch, but I could also include it in a comment here for clarity.

File: pkgs/development/compilers/gcc/default.nix (hunk 6 above)
Comment (reckenrode) [reply pkgs/development/compilers/gcc/default.nix:424]: While it’s not Darwin-specific, Darwin is the primary user of clang and was updated to clang 16 this staging cycle. This particular patch was prompted by potential issues I noticed while reviewing `config.log`, but I wasn’t at the point of being able to build GCC at the time, so it may be that it’s not actually necessary. I’m reviewing and building the affected versions to confirm.

File: pkgs/development/compilers/gcc/patches/default.nix (hunk 9 above)
Comment (reckenrode) [reply pkgs/development/compilers/gcc/patches/default.nix:305]: Which ones am I splitting off? If I can drop adb07a447b76325a54ee289fe181af91a64e6e04, then the only commit that doesn’t patches is f1f4d827cdbaae8bf7f5d1e8e8271c4adefe91b4. Wouldn’t it make more sense to split that one off and retarget this PR to master?

**Comment by reckenrode** [comment]
I’ve updated my branch based on the feedback I’ve received and force pushed it. I also retargeted master since this shouldn’t cause significant rebuilds. (It was targeting staging-next because it’s caused by the clang 16 update.)

**Comment by reckenrode** [comment]
I see `xgcc` processes in `Activity Monitor.app`, so I guess those failures weren’t important enough to prevent clang 16 from being able to build a functioning stage 1 GCC. It’s almost 2:30 AM here, so I need to get some sleep. I’ll let this build and one against master (using clang 11) run and check on it in the morning.

**Comment by amjoseph-nixpkgs** [comment]
> > I still don't understand why that was deleted if this emergency PR is just trying to fix Darwin.
> 
> It wasn’t deleted. It was moved. If I added the condition for clang 16 after the existing one, `NIX_CFLAGS_COMPILE` wouldn’t have `-Wno-unused-command-line-argument` when building gfortran7 on a clang 16 platform like Darwin.

You moved it into a `if` block triggered only by `!atLeast11 && stdenv.cc.isClang`.  So for any build which is `is7` but doesn't trigger those conditions, the derivation will now contain `NIX_CFLAGS_COMPILE=null` whereas before it contained `NIX_CFLAGS_COMPILE=""`.  That's why you're getting non-Darwin rebuilds.

I appreciate the desire to do some refactoring while you're at it, but for an urgent emergency-fix PR like this it's better to prioritize keeping the changes focused and limited to only touching or adding `if (stdenv.isDarwin) { ... }` blocks.

This is likely our last merge of `staging-next` before the release, so if we screw up here there is no second chance.

File: pkgs/development/compilers/gcc/patches/default.nix (hunk 7 above)
Comment (amjoseph-nixpkgs) [reply pkgs/development/compilers/gcc/patches/default.nix:198]: > This technically isn’t a Darwin-specific problem even if it’s primarily affecting x86_64-darwin.

Okay but only Darwin needs this super-urgently.

Please consider committing my suggestion, and sending another PR to undo it after the branch-off.

Comment (amjoseph-nixpkgs) [reply pkgs/development/compilers/gcc/patches/default.nix:198]: > It’s almost 2:30 AM here, so I need to get some sleep. 

In light of this, I'm going to commit my minor suggestion above and merge the result (you've addressed all of my concerns and this is now based on `master`).

It should be merged back to `staging-next` by the time you wake up.

Please don't hesitate to open a PR which undoes my suggested change; as long as somebody who knows using-clang-on-linux (I don't, @qyliss or @Ericson2314 probably do) signs off on it it's fine with me.

**Review by ghost — APPROVED** [review]
Thanks for getting all of this working, and for responding to my review so quickly!

> It’s almost 2:30 AM here, so I need to get some sleep. 

In light of this, I committed my minor suggestion above and will merge this as soon as CI passes.  It should be merged back to `staging-next` by the time you wake up.

File: pkgs/development/compilers/gcc/default.nix (hunk 4 above)
Comment (amjoseph-nixpkgs) [reply pkgs/development/compilers/gcc/default.nix]: Yeah; unfortunately `drv.NIX_CFLAGS_COMPILE=""` and `!(drv?NIX_CFLAGS_COMPILE)` hash differently.

**Comment by reckenrode** [comment]
> You moved it into a if block triggered only by !atLeast11 && stdenv.cc.isClang. So for any build which is is7 but doesn't trigger those conditions, the derivation will now contain NIX_CFLAGS_COMPILE=null whereas before it contained NIX_CFLAGS_COMPILE="". That's why you're getting non-Darwin rebuilds.

It wasn’t so much about refactoring as it was my not seeing a better way at the time to make sure I wasn’t replacing the value when both conditions are met. Well, thinking about it some more, there are ways to do it, but not needing the commit at all is better.

> This is likely our last merge of staging-next before the release, so if we screw up here there is no second chance.

Yeah. I’ve been trying to fix as much of the clang 16 fallout as I can before then. If it the update has to be reverted, that’s not a problem, but it would be preferable to get it in, so Darwin isn’t stuck on an old stack for another release. That would let us start working towards improving the SDK situation.

**Comment by reckenrode** [comment]
> It should be merged back to `staging-next` by the time you wake up.

It was. Thank you for the feedback and help getting this merged.
